stan/math/prim/functor/root_finder.hpp

```diff
@@ -0,0 +1,181 @@
+auto root_finder_halley_tol(FTuple&& f_tuple, const GuessScalar guess,
+                            const MinScalar min, const MaxScalar max,
+                            const int digits, std::uintmax_t& max_iter,
+                            Types&&... args) {
+  return root_finder_tol(
+      [](auto&&... args) {
+        return boost::math::tools::halley_iterate(args...);
+      },
+      std::forward<FTuple>(f_tuple), guess, min, max, digits, max_iter,
+      std::forward<Types>(args)...);
+}
+
+template <typename FTuple, typename GuessScalar, typename MinScalar,
+          typename MaxScalar, typename... Types>
+auto root_finder_newton_raphson_tol(FTuple&& f_tuple, const GuessScalar guess,
+                                    const MinScalar min, const MaxScalar max,
+                                    const int digits, std::uintmax_t& max_iter,
+                                    Types&&... args) {
+  return root_finder_tol(
+      [](auto&&... args) {
+        return boost::math::tools::newton_raphson_iterate(args...);
+      },
+      std::forward<FTuple>(f_tuple), guess, min, max, digits, max_iter,
+      std::forward<Types>(args)...);
+}
+
+template <typename FTuple, typename GuessScalar, typename MinScalar,
+          typename MaxScalar, typename... Types>
+auto root_finder_schroder_tol(FTuple&& f_tuple, const GuessScalar guess,
+                              const MinScalar min, const MaxScalar max,
+                              const int digits, std::uintmax_t& max_iter,
+                              Types&&... args) {
+  return root_finder_tol(
+      [](auto&&... args) {
+        return boost::math::tools::schroder_iterate(args...);
+      },
+      std::forward<FTuple>(f_tuple), guess, min, max, digits, max_iter,
+      std::forward<Types>(args)...);
+}
+
+/**
+ * Solve for root using Boost Halley method with default values for the
+ * tolerances
+ * @tparam FTuple A tuple holding functors whose signatures all match
+ *  `(GuessScalar g, Types&&... Args)`.
+ * @tparam GuessScalar Scalar type
+ * @tparam MinScalar Scalar type
+ * @tparam MaxScalar Scalar type
+ * @tparam Types Arg types to pass to functors in `f_tuple`
+ * @param f_tuple A tuple of functors to calculate the value and any derivates
+ * needed.
+ * @param guess An initial guess at the root value
+ * @param min The minimum possible value for the result, this is used as an
+ * initial lower bracket
+ * @param max The maximum possible value for the result, this is used as an
+ * initial upper bracket
+ * @param args Parameter pack of arguments to pass the the functors in `f_tuple`
+ */
+template <typename SolverFun, typename FTuple, typename GuessScalar,
+          typename MinScalar, typename MaxScalar, typename... Types>
+auto root_finder(SolverFun&& f_solver, FTuple&& f_tuple,
+                 const GuessScalar guess, const MinScalar min,
+                 const MaxScalar max, Types&&... args) {
+  constexpr int digits = 16;
```

Comment:
how was this default chosen? Maybe add one sentence about this choice in the doxygen doc.

Reply:
I need to update this to be in line with what boost's docs say 

> The value of digits is crucial to good performance of these functions, if it is set too high then at best you will get one extra (unnecessary) iteration, and at worst the last few steps will proceed by [bisection](https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/math_toolkit/roots/roots_noderiv/bisect.html). Remember that the returned value can never be more accurate than f(x) can be evaluated, and that if f(x) suffers from cancellation errors as it tends to zero then the computed steps will be effectively random. The value of digits should be set so that iteration terminates before this point: remember that for second and third order methods the number of correct digits in the result is increasing quite substantially with each iteration, digits should be set by experiment so that the final iteration just takes the next value into the zone where f(x) becomes inaccurate. A good starting point for digits would be 0.6*D for Newton and 0.4*D for Halley or Shröder iteration, where D is std::numeric_limits<T>::digits.

https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/math_toolkit/roots/roots_deriv.html